### DISCUSSION TITLE
Feature Request: Ability to load python udfs from SQL

### DISCUSSION THREAD

**Discussion by PadenZach** [duckdb/duckdb#12527] [Ideas]
My team currently uses dbt-duckdb significantly, one of the main reasons for doing this is the great python udf support duckdb currently has.

However; This makes it so that we currently need to:
- Create a method to load python udfs during `dbt run` (this is actually handled by the awesome dbt-duckdb adapter! )
- Create a way to load python extensions for ad-hoc/debug use cases.

For example I may have a python udf such as:

```python
def foobar(baz:int)-> str:
    return 'some string after python logic'

duckdb.create_function('foobar', foobar)
duckdb.sql(""" SELECT foobar(5);""")
```

Ideally, we can push this into a more sql world and support some functionality such as:
```sql
LOAD python_udf_loader;
SET python_path = '../python.exe'; -- or default to $(which python)
CALL load_python_udf('my_module.py', 'foobar');

SELECT foobar(5);
```

This would be great as it would then allow us to add python udfs to a ~/.duckdbrc file and improve the developer experience significantly.

### Related
https://github.com/duckdb/duckdb/issues/2849
https://github.com/MarkRoddy/duckdb-pytables 
https://github.com/duckdb/dbt-duckdb#configuring-dbt-duckdb-plugins
https://github.com/duckdb/duckdb/discussions/8684

**Comment by PadenZach** [comment]
I'm interested in figuring out what would actually be needed to get this working. Based on 2849 it seems like we'd need to have some pointer to the python udf, and essentially start an entire python process to run the udfs.

AFAIK (not a c++ dev so bare with me), it seems like the current python udf functionality may only work when ran from a python process. So in this case doing this the other way around may be more performant (this is essentially what the dbt-duckdb python udfs do.

What would be interesting is if there was maybe a way to launch a duckdb cli session from python, this would allow us to create a sort of debug process such as:

```python
import duckdb
from our_udf_package import register_all

if __name__ == "__main__":
    con = duckdb.connect()
    register_all(con)
    con.start_duckdb_cli()
```

```bash
> ./duckdb_harness.py
v1.0.0 1601d94f94
Enter ".help" for usage hints.
Connected to python session at pid 123
Use ".open FILENAME" to reopen on a persistent database.
D 
```

**Comment by soerenwolfers** [comment]
Which bit of the CLI experience do you prefer over running duckdb in your favorite Jupyter notebook host (JupyterLab, VSCode, Pycharm,...)?

With a tiny bit of cell magic to eliminate the need for triple quotes, I feel the experience in the latter becomes superior to the CLI

**Reply by PadenZach** [reply]
So the issue is that on our project not everyone is a python dev. Lots of folks are coming from heavy SQL or legacy DBA style backgrounds. where they are not comfortable writing/debugging python.

More importantly, having to run it all via python makes integrating python udfs into any other environments very difficult. (EG: dbeaver, harlequin, datagrip, or just plain 'ol bash)

**Reply by Tishj** [reply]
I think I can see a world where this would work, it would probably have to be a Catalog Extension, so it can be used by ATTACH.
Taking inspiration from MarkRoddy it would start an interpreter to be able to speak to python, override the `default_connection` of the `duckdb` module in the script that's attached

I'm not sure if this is something we'll pick up though, it doesn't sound like it has a lot of value to be frank, all of this can be done inside of python, instead of bringing python into the CLI as Soeren pointed out
